### PR TITLE
Clean up E0733 explanation

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @Dylan-DPC

### CODE REVIEW

File: src/librustc_error_codes/error_codes/E0733.md [hunk 1]
```diff
@@ -1,4 +1,6 @@
-Recursion in an `async fn` requires boxing. For example, this will not compile:
+An `async` function used recursion without boxing.
```
Comment (jyn514) [inline src/librustc_error_codes/error_codes/E0733.md:1]:
```suggestion
An [`async`] function used recursion without boxing.

[`async`]: https://doc.rust-lang.org/std/keyword.async.html
```

File: src/librustc_error_codes/error_codes/E0733.md [hunk 2]
```diff
@@ -8,8 +10,8 @@ async fn foo(n: usize) {
 }
 ```
 
-To achieve async recursion, the `async fn` needs to be desugared
-such that the `Future` is explicit in the return type:
+To achieve async recursion, the `async fn` needs to be desugared such that the
```
Comment (jyn514) [inline src/librustc_error_codes/error_codes/E0733.md:13]:
```suggestion
To perform async recursion, the `async fn` needs to be desugared such that the
```

**Review by jyn514 — COMMENTED** [review]
As a general note - this explains _how_ to fix the error, but not why it's an error in the first place. Do you know why? It's not obvious to me.

**Comment by GuillaumeGomez** [comment]
@jyn514 I assume that it's because of the stack but honestly, I'm not really sure.

**Comment by jyn514** [comment]
@bors r+ rollup=always

:shrug: mysteries of `async`

**Pull request merged**